File: Data Structures & Algorithms/minimum-window-with-characters/submission-1.cpp
class Solution {
public:
    string minWindow(string s, string t) {
        unordered_map<char, int> need, window;
        for(char ch: t) need[ch]++;
        int req=need.size();
        int have=0;
        int minLen=INT_MAX, l=0, start=0;
        for(int r=0;r<s.size();r++) {
            window[s[r]]++;
            if(need.count(s[r]) && need[s[r]]== window[s[r]]) have++;
            while(have == req) {
                // update the ans
                if(minLen > r-l+1) {
                    minLen = r-l+1;
                    start = l;
                }
                window[s[l]]--;
                if(need.count(s[l]) && need[s[l]] > window[s[l]]) have--;
                l++;
            }
        }
        return minLen==INT_MAX ? "" : s.substr(start, minLen);
    }
};

File: Data Structures & Algorithms/climbing-stairs/submission-1.py
class Solution:

    def count(self, n: int) -> int :
        if n <= 2:
            return n
        if self.steps[n] != -1:
            return self.steps[n]
        self.steps[n] = self.count(n-1) + self.count(n-2)
        return self.steps[n]

    def climbStairs(self, n: int) -> int:
        self.steps = [-1] * (n+1)
        return self.count(n)
